aten/src/ATen/native/cuda/MemoryAccess.cuh

```diff
@@ -0,0 +1,120 @@
+#pragma once
+
+#include <cstdint>
+#include <type_traits>
+#include <c10/util/Exception.h>
+
+// References:
+// https://devblogs.nvidia.com/cuda-pro-tip-increase-performance-with-vectorized-memory-access/
+
+namespace at { namespace native { namespace memory {
+
+// aligned vector generates vectorized load/store on CUDA
+template<typename scalar_t, int vec_size>
+struct alignas(sizeof(scalar_t) * vec_size) aligned_vector {
+  scalar_t val[vec_size];
+};
+
+template <
+  int num_threads_,        // number of threads in a block.
+  int thread_work_size_    // number of elements each block needs to handle.
+>
+struct policies {
+
+  struct common {
+    static constexpr int num_threads = num_threads_;
+    static constexpr int thread_work_size = thread_work_size_;
+    static constexpr int block_work_size = thread_work_size_ * num_threads_;
+  };
+
+  struct checked_unroll : public common {
```

Comment:
Can you do without the common struct? Will C++ let you declare the static constexpr ints as members of policies and use them from checked_unroll and vectorized without inheriting another struct?

Reply:
The reason for this is because it is used in `Loops.cuh` as

```C++
template<t..., typename policy_t>
__device__ void elementwise_kernel_helper(..., policy_t policy) {
  ...
  return_t results[policy_t::thread_work_size];
  arg_t args[policy_t::thread_work_size][nargs];
  ...
  #pragma unroll
  for (int i = 0; i < policy_t::thread_work_size; i++) {
  ...
}
```

`elementwise_kernel_helper` does not see the arguments passed to the container struct `policies`, we have to pass it to `policy_t`.